sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala

```diff
@@ -192,41 +253,63 @@ class StateSchemaCompatibilityChecker(
-      createSchemaFile(newStateSchemaList.sortBy(_.colFamilyName), stateSchemaVersion)
+      stateSchemaVersion: Int,
+      schemaEvolutionEnabled: Boolean): Boolean = {
+    val existingStateSchemaMap = readSchemaFiles()
+    val mostRecentColFamilies = getExistingKeyAndValueSchema().map(_.colFamilyName)
+    if (mostRecentColFamilies.isEmpty) {
+      // Initialize schemas with ID 0 when no existing schema
+      val initializedSchemas = newStateSchema.map(schema =>
+        schema.copy(keySchemaId = 0, valueSchemaId = 0)
+      )
+      createSchemaFile(initializedSchemas.sortBy(_.colFamilyName), stateSchemaVersion)
       true
     } else {
-      // validate if the new schema is compatible with the existing schema
-      val existingSchemaMap = existingStateSchemaList.map { schema =>
-        schema.colFamilyName -> schema
-      }.toMap
-      // For each new state variable, we want to compare it to the old state variable
-      // schema with the same name
-      newStateSchemaList.foreach { newSchema =>
-        existingSchemaMap.get(newSchema.colFamilyName).foreach { existingStateSchema =>
-          check(existingStateSchema, newSchema, ignoreValueSchema)
-        }
+
+      // Process each new schema and track if any have evolved
+      val (evolvedSchemas, hasEvolutions) = newStateSchema.foldLeft(
+        (List.empty[StateStoreColFamilySchema], false)) {
+        case ((schemas, evolved), newSchema) =>
+          existingStateSchemaMap.get(newSchema.colFamilyName) match {
+            case Some(existingSchemas) =>
+              val (updatedSchema, hasEvolved) = check(
+                existingSchemas, newSchema, ignoreValueSchema, schemaEvolutionEnabled)
+              (updatedSchema :: schemas, evolved || hasEvolved)
+            case None =>
+              // New column family - initialize with schema ID 0
+              val newSchemaWithIds = newSchema.copy(keySchemaId = 0, valueSchemaId = 0)
+              (newSchemaWithIds :: schemas, true)
+          }
       }
-      val colFamiliesAddedOrRemoved =
-        (newStateSchemaList.map(_.colFamilyName).toSet != existingSchemaMap.keySet)
-      if (stateSchemaVersion == SCHEMA_FORMAT_V3 && colFamiliesAddedOrRemoved) {
-        createSchemaFile(newStateSchemaList.sortBy(_.colFamilyName), stateSchemaVersion)
+
+      val newColFamilies = newStateSchema.map(_.colFamilyName).toSet
+      val oldColFamilies = mostRecentColFamilies.toSet
+      val colFamiliesAddedOrRemoved = newColFamilies != oldColFamilies
+      val newSchemaFileWritten = hasEvolutions || colFamiliesAddedOrRemoved
+
+      if (oldSchemaFilePaths.size == conf.streamingStateSchemaFilesThreshold &&
+        colFamiliesAddedOrRemoved) {
+        throw StateStoreErrors.streamingStateSchemaFilesThresholdExceeded(
+          oldSchemaFilePaths.size + 1,
+          conf.streamingStateSchemaFilesThreshold,
+          // need to compute symmetric diff between col family list
+          (newColFamilies.diff(oldColFamilies) ++
```

Comment:
Why not be more informative? I think it's lot better to provide the information of prior column families vs new column families, or differentiate the representation between cfs being added vs cfs being removed, e.g. `+cf1, +cf2, -cf3, -cf4` / or `added: [cf1, cf2], removed: [cf3, cf4]`.